### PR TITLE
Improve User Details Page and Company Page's User Dropdown Performances

### PR DESCRIPTION
**Pull request by calvinlu3**
This fixes https://github.com/oncokb/oncokb/issues/2342

- Cache `getAllManagedUsers` method in `UserService.java`
- Update `usersToUserDTOs` method to fetch user + user details in bulk
- Update `getNonCompanyUserEmails` method in `UserService.java` to use custom query
   - The original query takes about 7-8 seconds, but the custom query takes ~50ms.

### CODE REVIEW

**Comment by calvinlu3** [comment]
@zhx828 , I had to put `findAllUsersWithUserDetailsByUsersIn(...)` in `UserDetailsService` instead of `UserService` due to circular dependencies.

File: src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java [hunk 1]
```diff
@@ -84,4 +85,17 @@ public Optional<UserDetailsDTO> findByUserIsCurrentUser() {
...
+
+    @Transactional(readOnly = true)
+    public List<UserAndUserDetailsPair> findAllUsersWithUserDetailsByUsersIn(List<User> users) {
+        List<Object[]> usersWithDetails = userDetailsRepository.findAllUsersWithUserDetailsByUsersIn(users);
+        return usersWithDetails
+            .stream()
+            .map(u -> {
+                User user = (User) u[0];
+                UserDetails userDetails = (UserDetails) u[1];
+                return new UserAndUserDetailsPair(user, userDetails);
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java:97]: I just realize that isn't this UserDTO? Let's return the UserDTO directly?

**Review by zhx828 — CHANGES_REQUESTED** [review]
We need to address the circular dependency issue in the future. The mapper files probably should leave the repository/service out.

File: src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java (hunk 1 above)
Comment (calvinlu3) [reply src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java:97]: Yeah it is very similar. The reason is that [we add the companyDTO](https://github.com/oncokb/oncokb-public/blob/master/src/main/java/org/mskcc/cbio/oncokb/service/mapper/UserMapper.java#L40) when mapping the user to userDTO. If I used the UserDTO, then I cannot map to a CompanyDTO.

Comment (zhx828) [reply src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java:97]: Are you adding it here? https://github.com/oncokb/oncokb-public/pull/896/files#diff-ccdbff722436ab7c2f7d0a2d7ecc10a40b0b25de777c602ff46dce2e588d8009R41

Comment (calvinlu3) [reply src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java:97]: Yeah, I am adding the companyDTO at that line. If I returned a UserDTO, then I'm not able to add that line in the UserMapper.

Comment (zhx828) [reply src/main/java/org/mskcc/cbio/oncokb/service/impl/UserDetailsServiceImpl.java:97]: Should we add a new method in UserDTO, a copy of https://github.com/oncokb/oncokb-public/blob/master/src/main/java/org/mskcc/cbio/oncokb/service/mapper/UserMapper.java#L36 but accepting UserDetails as second parameter?

File: src/main/java/org/mskcc/cbio/oncokb/service/UserService.java [hunk 2]
```diff
@@ -508,8 +507,24 @@ public void changePassword(String currentClearTextPassword, String newPassword)
...
+                User user = (User) u[0];
+                UserDetails userDetails = (UserDetails) u[1];
+                return userMapper.userToUserDTO(user, userDetails);
+            })
+            .collect(Collectors.toList());
+    }
+
+    @Transactional(readOnly = true)
+    @Cacheable(cacheResolver = "userCacheResolver", key = "#root.methodName")
     public Page<UserDTO> getAllManagedUsers(Pageable pageable) {
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/service/UserService.java:524]: Could you do me a favor checking whether we can get rid of the threshold https://github.com/oncokb/oncokb-public/blob/master/src/main/webapp/app/pages/userManagement/UserDetailsPage.tsx#L78? So the pageable will be null here I assume? If not, let's change that to a bigger size? 10K? @calvinlu3

Comment (calvinlu3) [reply src/main/java/org/mskcc/cbio/oncokb/service/UserService.java:524]: The Pageable object populates with default parameters when we get rid of the threshold. The default size is 20.
I'm increasing the page size to 10k.

**Review by zhx828 — APPROVED** [review]

**Pull request merged**